### PR TITLE
feat: dedupe results by unique fingerprint

### PR DESCRIPTION
**Pull request by marbemac**
- [x] merge #839 first, this PR was branched off of that work

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [x] Yes

Duplicate lint results are no longer reported.

### CODE REVIEW

File: docs/guides/javascript.md [hunk 1]
```diff
@@ -199,3 +202,28 @@ const spectral = new Spectral({ resolver: customFileResolver });
 The custom resolver we've just created will resolve all remote file refs relatively to the current working directory.
 
 More on that can be found in the [json-ref-resolver repo](https://github.com/stoplightio/json-ref-resolver).
+
+### Using custom de-duplication strategy
+
+By default, Spectral will de-duplicate results based on the result code and document location. You can customize this
```
Comment (P0lip) [inline docs/guides/javascript.md]: How about using deduplicateResults function I implemented back in a day and nulltoken vastly improved recently? I think we just don't call it in a run method. We just call it in the CLI mode.

Comment (marbemac) [reply docs/guides/javascript.md]: We could - however the fingerprint approach is more elegant IMHO, and easy to customize - does deduplicateResults cover things the default compute fingerprint func in this PR does not?

Comment (nulltoken) [reply docs/guides/javascript.md]: @marbemac internally `deduplicateResults` relies on a sorting comparer. As such, the deduplicated results are also... sorted ;-) This allows the linter to always provide a predictable and stable output in the test harness and cli output.

However, I tend to agree with @P0lip that having two "somewhat-but-not-completely-similar" functions might not be ideal. How would you feel about making deduplicateResults leverage this fingerprint based approach and then eventually sort the resulting array of results?

**Comment by marbemac** [comment]
Alright, combined deduplicateResults and cleaned up tests since things are always sorted now.

The karma tests are not working for whatever reason, will need some help @P0lip to get tests across finish line. No rush.

**Comment by P0lip** [comment]
Karma tests fail since the entire `src/formatters` directory is excluded, as the code there is meant to be used by CLI. I'll move `sortResults` function as it's used by the core Spectral part now.

File: src/utils/prepareResults.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+import { compareResults } from '../formatters/utils/sortResults';
+import { IRuleResult } from '../types';
+
+export type ComputeFingerprintFunc = (rule: IRuleResult, hash: (val: string) => string) => string;
+
+export const defaultComputeResultFingerprint: ComputeFingerprintFunc = (rule, hash) => {
+  let id = String(rule.code);
+
+  if (rule.path && rule.path.length) {
```
Comment (P0lip) [inline src/utils/prepareResults.ts]:
```suggestion
  if (rule.path.length) {
```
`IRuleResult` enforces `path` to be defined.

File: src/spectral.ts [hunk 3]
```diff
@@ -111,9 +115,11 @@ export class Spectral {
       }
     }
 
+    validationResults = validationResults.concat(runRules(resolved, this.rules, this.functions));
```
Comment (P0lip) [inline src/spectral.ts]: Any reason not to do the following?
```ts
prepareResults(
  [...validationResults, ...runRules(resolved, this.rules, this.functions)],
  this._computeFingerprint,
);
```

**Pull request merged**